docs/ClassBasedPlugins.md

```diff
@@ -0,0 +1,131 @@
+
+# Class based plugins
+
+> Experimental
+
+**Note**: It is required to enable the `experimentalDecorators` compiler option to be able to use the typescript decorators.
+
+```ts
+class MyPlugin {
+  constructor (options) {
+    // ...
+  }
+
+  @Get('/')
+  async route (request, reply) {
+    return 'hello world'
+  }
+}
+
+fastify.register(new MyPlugin({ /* plugin options */ }))
+```
+
+## Routes
+
+Routes can be defined with the method decorators.
+
+There is a decorator for each http method supported by fastify: `Get`, `Post`, `Delete`, `Put`, `Head`, `Patch` and `Options`
+
+Furthermore there is a `All` decorator that allows to add the same handler to all the supported methods.
+
+The first argument of the method decorators is the url.
+
+```ts
+import { Get, Post } from 'fastify'
+
+class UserController {
+  @Get('/user/:id')
+  async getUser (request, reply) {
+    // ...
+  }
+
+  @Post('/user/')
+  async createUser (request, reply) {
+    // ...
+  }
+}
+```
+
+The second argument of the method decorators are options (for available options see [here](./Routes.md)):
+
+```ts
+import { Get } from 'fastify'
+
+const schema = {
+  querystring: {
+    name: { type: 'string' },
+    excitement: { type: 'integer' }
+  },
+  response: {
+    200: {
+      type: 'object',
+      properties: {
+        hello: { type: 'string' }
+      }
+    }
+  }
+}
+
+class MyPlugin {
+  @Get('/', { schema })
+  async route (request, reply) {
+    return { hello: 'world' }
+  }
+}
+```
+
+## Hooks
+
+```ts
+import { Hook } from 'fastify'
+
+class MyPlugin {
+  @Hook('onRequest')
+  hook (request, reply, next) {
+    // ...
+    next()
+  }
+
+  @Hook('onRequest')
+  async asyncHook (request, reply) {
+    // ...
+  }
+}
+```
+
+## Decoration of request, reply or instance
+
+```ts
+import { DecorateInstance } from 'fastify'
+
+class MyPlugin {
+  @DecorateInstance()
```

Comment:
We would need to add a decorator on the class itself to enable/disable encapsulation, i.e. setting https://github.com/fastify/fastify-plugin/blob/966248ce75eebe58727aaa5c676d410b4641d236/index.js#L16

Reply:
I agree. Shall we add this decorator in `fastify-plugin`? It could provide the other features of `fastify-plugin` as a decorator, too.

Reply:
I opened a PR (fastify/fastify-plugin#73) in the fastify-plugin repository that adds such a decorator.